### PR TITLE
Empty custom header default

### PR DESCRIPTION
**Pull request by paul-r-gall**
Commit Message: [router] add additional default option for un-resolvable dynamic value.
Additional Description: Add an option so that config can specify whether to either append an empty string or just not add the custom header when a dynamic value cannot be resolved. See https://github.com/envoyproxy/envoy/issues/21265.
Risk Level: Low, no default behavior change.
Testing: Unit tests

### CODE REVIEW

File: api/envoy/config/core/v3/base.proto [hunk 1]
```diff
@@ -357,6 +362,8 @@ message HeaderValueOption {
   // [#not-implemented-hide:] Describes the action taken to append/overwrite the given value for an existing header
   // or to only add this header if it's absent. Value defaults to :ref:`APPEND_IF_EXISTS_OR_ADD<envoy_v3_api_enum_value_config.core.v3.HeaderValueOption.HeaderAppendAction.APPEND_IF_EXISTS_OR_ADD>`.
   HeaderAppendAction append_action = 3 [(validate.rules).enum = {defined_only: true}];
+
+  HeaderValueEmptyAction empty_action = 4 [(validate.rules).enum = {defined_only: true}];
```
[the remark this answers is not in the document]
Comment (paul-r-gall) [reply api/envoy/config/core/v3/base.proto:366]: The behavior I was imagining is as follows:

if (value is empty and empty_action = NO_HEADER) --> do nothing
if (value is empty and empty_action = EMPTY STRING) --> take `append_action` with the empty string. 

So these two enums are somewhat orthogonal.

File: source/common/router/header_parser.h [hunk 2]
```diff
@@ -70,6 +70,7 @@ class HeaderParser : public Http::HeaderEvaluator {
   struct HeadersToAddEntry {
     HeaderFormatterPtr formatter_;
     const std::string original_value_;
+    const bool add_if_empty_ = false;
```
Comment (adisuissa) [inline source/common/router/header_parser.h:73]: If this is const and initialized here, should it be constexpr?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!

I've found a few related issues/PRs: #3971, #7521 and #7564 (and specifically this [comment](https://github.com/envoyproxy/envoy/issues/7521#issuecomment-829441928))
IIUC the empty header values are allowed by the spec, but seem to be discouraged.
Maybe @mattklein123 @alyssawilk @PiotrSikora know more about this.

File: source/common/router/header_parser.h (hunk 2 above)
Comment (paul-r-gall) [reply source/common/router/header_parser.h:73]: Non-static data members can't be constexpr.

**Comment by paul-r-gall** [comment]
https://github.com/envoyproxy/envoy/issues/21265#issuecomment-1126183746 <-- Matt already commented on the issue.

**Comment by alyssawilk** [comment]
yeah I think many proxies (possibly even Envoy?) will remove the empty headers on the next hop, but as long as we have protocol integration tests verifying that Envoy will serialize it for all 3 codecs, and it's an explicit configuration option (folks know they're turning it up) I think it's totally fine.

File: api/envoy/config/core/v3/base.proto [hunk 3]
```diff
@@ -357,6 +357,8 @@ message HeaderValueOption {
   // [#not-implemented-hide:] Describes the action taken to append/overwrite the given value for an existing header
   // or to only add this header if it's absent. Value defaults to :ref:`APPEND_IF_EXISTS_OR_ADD<envoy_v3_api_enum_value_config.core.v3.HeaderValueOption.HeaderAppendAction.APPEND_IF_EXISTS_OR_ADD>`.
   HeaderAppendAction append_action = 3 [(validate.rules).enum = {defined_only: true}];
+
```
Comment (adisuissa) [inline api/envoy/config/core/v3/base.proto:360]: Please add a comment.

File: source/common/router/header_parser.h (hunk 2 above)
Comment (adisuissa) [reply source/common/router/header_parser.h:73]: I might be missing something here, but isn't this always false (i.e., initialized here, and doesn't change)?
If so, why is it per-entry data-member?

**Review by adisuissa — COMMENTED** [review]
Thanks!
Per Alyssa's comment I suggest to add some integration test that validates that this header will be kept all the way through (in requests and responses).

File: api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto [hunk 4]
```diff
@@ -391,8 +391,7 @@ message FilterStateRule {
 
   // A map of string keys to requirements. The string key is the string value
   // in the FilterState with the name specified in the *name* field above.
-  map<string, JwtRequirement>
-  requires = 3;
+  map<string, JwtRequirement> requires = 3;
```
Comment (adisuissa) [inline api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:394]: Change possibly caused by wrong tooling.

**Review by adisuissa — COMMENTED** [review]
Overall LGTM.
/assign-from @envoyproxy/senior-maintainers for next round of reviews

**Comment by adisuissa** [comment]
Please also add release notes at `changelogs/current.yaml`.

File: source/common/router/header_parser.h (hunk 2 above)
Comment (paul-r-gall) [reply source/common/router/header_parser.h:73]: false is just the default value, it's still possible to set the value using the default constructor -- i.e. HeadersToAddEntry{std::move(formatter), "orig_val", true};

Comment (adisuissa) [reply source/common/router/header_parser.h:73]: Ah, thanks. I've missed the struct initializer.

**Review by adisuissa — COMMENTED** [review]
Thanks, LGTM.
Seems that merge main is needed.

**Review by adisuissa — APPROVED** [review]
LGTM thanks!
/lgtm api

**Review by lizan — APPROVED** [review]

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**